File: cc/leetcode/array/1465_max_area.cc
class Solution {
public:
    int maxArea(int h, int w, vector<int>& horizontalCuts, vector<int>& verticalCuts) {
        sort(horizontalCuts.begin(), horizontalCuts.end());
        horizontalCuts.push_back(h);

        sort(verticalCuts.begin(), verticalCuts.end());
        verticalCuts.push_back(w);
        
        int max_h=horizontalCuts[0], max_w=verticalCuts[0];
        for (int i = 1; i < horizontalCuts.size(); i++)
            max_h = max(max_h, horizontalCuts[i]-horizontalCuts[i-1]);
        
        for (int i = 1; i < verticalCuts.size(); i++)
            max_w = max(max_w, verticalCuts[i]-verticalCuts[i-1]);
        
        long result = static_cast<long>(max_h) * static_cast<long>(max_w);
        long MOD = static_cast<long>(pow(10, 9) + 7);
        result %= MOD;
        return int(result);

    }
};

File: python/leetcode/array/1465_max_area.py
"""
1465. Maximum Area of a Piece of Cake After Horizontal and Vertical Cuts (Medium)

Given a rectangular cake with height h and width w, and two arrays of integers horizontalCuts and verticalCuts where horizontalCuts[i] is the distance from the top of the rectangular cake to the ith horizontal cut and similarly, verticalCuts[j] is the distance from the left of the rectangular cake to the jth vertical cut.

Return the maximum area of a piece of cake after you cut at each horizontal and vertical position provided in the arrays horizontalCuts and verticalCuts. Since the answer can be a huge number, return this modulo 10^9 + 7.

 

Example 1:



Input: h = 5, w = 4, horizontalCuts = [1,2,4], verticalCuts = [1,3]
Output: 4 
Explanation: The figure above represents the given rectangular cake. Red lines are the horizontal and vertical cuts. After you cut the cake, the green piece of cake has the maximum area.
Example 2:



Input: h = 5, w = 4, horizontalCuts = [3,1], verticalCuts = [1]
Output: 6
Explanation: The figure above represents the given rectangular cake. Red lines are the horizontal and vertical cuts. After you cut the cake, the green and yellow pieces of cake have the maximum area.
Example 3:

Input: h = 5, w = 4, horizontalCuts = [3], verticalCuts = [3]
Output: 9
 

Constraints:

2 <= h, w <= 10^9
1 <= horizontalCuts.length < min(h, 10^5)
1 <= verticalCuts.length < min(w, 10^5)
1 <= horizontalCuts[i] < h
1 <= verticalCuts[i] < w
It is guaranteed that all elements in horizontalCuts are distinct.
It is guaranteed that all elements in verticalCuts are distinct.
"""


class Solution(object):
    def maxArea(self, h, w, horizontalCuts, verticalCuts):
        """
        :type h: int
        :type w: int
        :type horizontalCuts: List[int]
        :type verticalCuts: List[int]
        :rtype: int
        """
        horizontalCuts.sort()
        verticalCuts.sort()
        MOD = 10 ** 9 + 7
        
        max_hori = horizontalCuts[0]
        horizontalCuts.append(h)
        for i, item in enumerate(horizontalCuts):
            # print(max_hori, item, horizontalCuts[i-1])
            if i > 0:
                max_hori = max(max_hori, item - horizontalCuts[i-1])
            # print(max_hori, item, horizontalCuts[i-1])
                
        max_vert = verticalCuts[0]
        verticalCuts.append(w)
        for i, item in enumerate(verticalCuts):
            if i > 0:
                max_vert = max(max_vert, item - verticalCuts[i-1])
                
        res = max_hori * max_vert % MOD
        return res
